google/cloud/firestore_v1/bulk_writer.py

```diff
@@ -181,25 +252,51 @@ class BulkWriter(AsyncBulkWriterMixin):
 
     def __init__(
         self,
-        client: Optional[BaseClient] = None,
+        client: Optional["BaseClient"] = None,
         options: Optional["BulkWriterOptions"] = None,
     ):
-        self._client = client
+        # Because `BulkWriter` instances are all synchronous/blocking on the
+        # main thread, instead using other threads for asynchrony, the
+        # `AsyncClient`'s various methods that return Futures are incompatible.
+        # `BulkWriter` parallelizes all of its network I/O without the developer
+        # having to worry about awaiting async methods, so we must convert an
+        # AsynClient into a plain Client instance.
+        self._client = (
+            client._to_sync_copy() if type(client).__name__ == "AsyncClient" else client
+        )
         self._options = options or BulkWriterOptions()
         self._send_mode = self._options.mode
-        # Redundantly set this variable for IDE type hints
-        self._batch: BulkWriteBatch = self._reset_batch()
+
+        self._operations: List[BulkWriterOperation]
+        # List of the `_document_path` attribute for each DocumentReference
+        # contained in the current `self._operations`. This is reset every time
+        # `self._operations` is reset.
+        self._operations_document_paths: List[BaseDocumentReference]
+        self._reset_operations()
+
+        # List of all `BulkWriterOperation` objects that are waiting to be retried.
+        # Each such object is wrapped in an `OperationRetry` object which pairs
+        # the raw operation with the `datetime` of its next scheduled attempt.
+        # `self._retries` must always remain sorted for efficient reads, so it is
+        # required to only ever add elements via `bisect.insort`.
+        self._retries: collections.deque["OperationRetry"] = collections.deque([])
+
         self._queued_batches = collections.deque([])
         self._is_open: bool = True
-        self._is_sending: bool = False
 
-        self._success_callback: Optional[
-            Callable[[BaseDocumentReference, WriteResult, "BulkWriter"], NoReturn]
-        ] = None
-        self._batch_callback: Optional[
-            Callable[[BulkWriteBatch, BatchWriteResponse, "BulkWriter"], NoReturn]
-        ] = None
-        self._error_callback: Optional[Callable] = None
+        # Stores all currently pending futures for the purposes of the `flush`
```

Comment:
This comment seems a bit odd to be attached to an empty list. Is there something else going on